### PR TITLE
[MU4] Extensions page

### PR DESCRIPTION
**Pull request by Eism**
Peek 2020-07-29 12-34

### CODE REVIEW

File: framework/CMakeLists.txt [hunk 1]
```diff
@@ -6,6 +6,7 @@ add_subdirectory(actions)
 add_subdirectory(shortcuts)
 add_subdirectory(workspace)
 add_subdirectory(network)
+add_subdirectory(system)
```
Comment (igorkorsukov) [inline framework/CMakeLists.txt:9]: What else can you put into a systems module?
I am confused that the modules of the network, system are very small
Maybe they should be merged?

Comment (igorkorsukov) [reply framework/CMakeLists.txt:9]: Or maybe fsoperation can be moved to the global?

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    file.close();
+
+    result.ret = make_ret(Err::NoError);
+    result.val = extensionArchivePath;
+    return result;
+}
+
+Ret ExtensionsController::removeExtension(const QString& extensionCode) const
+{
+    QString extensionPath = configuration()->extensionsSharePath() + "/" + extensionCode;
```
Comment (igorkorsukov) [inline mu4/extensions/internal/extensionscontroller.cpp:294]: I see it at least the second time
Better to do this:
```
QString extensionPath = configuration()->makeExtensionPath(extensionCode);
```
And all these things are better done this way.
This eliminates duplication, we can easily see which paths are where.

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        return make_ret(Err::ErrorRemoveExtensionDirectory);
+    }
+
+    return make_ret(Err::NoError);
+}
+
+Extension::ExtensionTypes ExtensionsController::extensionTypes(const QString& extensionCode) const
+{
+    Extension::ExtensionTypes result;
+    QString workspacesPath(configuration()->extensionsSharePath() + "/" + extensionCode + "/workspaces");
```
Comment (igorkorsukov) [inline mu4/extensions/internal/extensionscontroller.cpp:306]:
```
QString workspacesPath = configuration()->extensionWorkspacesPath(extensionCode);
```

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+RetVal<QString> ExtensionsController::downloadExtension(const QString& extensionCode,
+                                                        async::Channel<ExtensionProgress>& progressChannel) const
+{
+    RetVal<QString> result;
+
+    ValCh<ExtensionsHash> extensions = configuration()->extensions();
+    QString fileName = extensions.val.value(extensionCode).fileName;
+
+    QString extensionArchivePath = configuration()->extensionsDataPath() + "/" + fileName;
```
Comment (igorkorsukov) [inline mu4/extensions/internal/extensionscontroller.cpp:263]: QString extensionArchivePath = configuration()->extensionArchivePath(fileName);

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+        result.val.insert(key, extension);
+    }
+
+    return result;
+}
+
+bool ExtensionsController::isExtensionExists(const QString& extensionCode) const
+{
+    return fsOperation()->exists(configuration()->extensionsSharePath() + "/" + extensionCode);
```
Comment (igorkorsukov) [inline mu4/extensions/internal/extensionscontroller.cpp:225]:
```
configuration()->extensionPath(extensionCode);
```

File: framework/system/systemmodule.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+#define MU_FRAMEWORK_SYSTEMMODULE_H
+
+#include "modularity/imodulesetup.h"
+
+namespace mu {
+namespace framework {
+class SystemModule : public IModuleSetup
+{
+public:
+
```
Comment (RomanPudashkin) [inline framework/system/systemmodule.h:29]: Please remove the empty line

File: framework/uicomponents/qml/MuseScore/UiComponents/PopupPanel.qml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    function hide() {
+        visible = false
+
+        closed()
+    }
+
+    Loader {
+        id: loader
+
```
Comment (RomanPudashkin) [inline framework/uicomponents/qml/MuseScore/UiComponents/PopupPanel.qml:35]: readonly property int sideMargin: 68
anchors.left: sideMargin
anchors.right: sideMargin?

File: framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    states: [
+        State {
+            name: "PRESSED"
+            when: clickableArea.pressed
+
+            PropertyChanges {
+                target: backgroundRect
+                opacity: ui.theme.buttonOpacityHit
+                border.color: "#25000000"
```
Comment (RomanPudashkin) [inline framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml:102]: We need to take colors from ui.theme

File: framework/uicomponents/qml/MuseScore/UiComponents/StyledTabButton.qml [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import QtQuick 2.12
+import QtQuick.Controls 2.12
+import MuseScore.UiComponents 1.0
+
+TabButton {
+    id: root
+
+    property int sideMargin: 0
```
Comment (RomanPudashkin) [inline framework/uicomponents/qml/MuseScore/UiComponents/StyledTabButton.qml:8]: Can we remove this?

Comment (Eism) [reply framework/uicomponents/qml/MuseScore/UiComponents/StyledTabButton.qml:8]: No, by default, the buttons in the toolBar are positioned without spacing. This is an alternative to spacing.

File: framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml (hunk 8 above)
Comment (Eism) [reply framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml:102]: I did like in FlatButton.

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+
+void ExtensionsController::init()
+{
+    fsOperation()->makePath(configuration()->extensionsSharePath());
+    fsOperation()->makePath(configuration()->extensionsDataPath());
+
+    refreshExtensions();
+}
+
+Ret ExtensionsController::refreshExtensions()
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionscontroller.cpp:40]: Function seems to be too difficult. Can we divide it?

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+        }
+
+        ExtensionsHash extensionHash = this->extensions().val;
+
+        extensionHash[extensionCode].status = ExtensionStatus::Installed;
+        extensionHash[extensionCode].types = extensionTypes(extensionCode);
+
+        Ret updateConfigRet = configuration()->setExtensions(extensionHash);
+        if (!updateConfigRet) {
+            return; // TODO
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionscontroller.cpp:115]: Please fix that todo

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+        }
+
+        ExtensionsHash extensionHash = extensions().val;
+
+        extensionHash[extensionCode].status = ExtensionStatus::Installed;
+        extensionHash[extensionCode].types = extensionTypes(extensionCode);
+
+        Ret updateConfigRet = configuration()->setExtensions(extensionHash);
+        if (!updateConfigRet) {
+            return; // TODO
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionscontroller.cpp:145]: Please fix that todo

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+    }
+
+    return result;
+}
+
+void ExtensionsController::th_install(const QString& extensionCode,
+                                      async::Channel<ExtensionProgress> progressChannel,
+                                      std::function<void(const QString&, const Ret&)> callback)
+{
+    progressChannel.send(ExtensionProgress(qtrc("extensions", "Analysing..."), true));
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionscontroller.cpp:317]: static const QString ANALYSING_STATUS = qtrc("extensions", "Analysing...");

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+    RetVal<QString> result;
+
+    QString extensionArchivePath = configuration()->extensionArchivePath(extensionCode);
+
+    QBuffer buff;
+    INetworkManagerPtr networkManagerPtr = networkManagerCreator()->makeNetworkManager();
+
+    async::Channel<Progress> downloadChannel = networkManagerPtr->downloadProgressChannel();
+    downloadChannel.onReceive(new deto::async::Asyncable(), [&progressChannel](const Progress& progress) {
+        progressChannel.send(ExtensionProgress(qtrc("extensions", "Downloading..."), progress.current,
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionscontroller.cpp:267]: static const QString DOWNLOADING_STATUS = qtrc("extensions", "Downloading...");

File: mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+            clip: true
+
+            Rectangle {
+                anchors.fill: parent
+                anchors.rightMargin: -12 // NOTE: for a rounded corner on the left side only
+                radius: root.radius
+                border.width: root.border.width
+                border.color: root.border.color
+
+                color: "#595959" // TODO
```
Comment (RomanPudashkin) [inline mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml:36]: Please fix that todo

File: mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+
+                    text: qsTrc("extensions", "Placeholder")
+                    font.pixelSize: 30
+                }
+            }
+        }
+
+        Item {
+            Layout.fillHeight: true
+            Layout.fillWidth: true
```
Comment (RomanPudashkin) [inline mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml:49]: readonly property int margin: 58

and use that in nameLabel and descriptionLabel

File: mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+                maximumLineCount: 2
+
+                font.pixelSize: 22
+                font.bold: true
+            }
+
+            StyledTextLabel {
+                id: descriptionLabel
+
+                anchors.top: nameLabel.bottom
```
Comment (RomanPudashkin) [inline mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml:71]: Maybe change parent to Column?

File: mu4/extensions/qml/MuseScore/Extensions/ExtensionsContent.qml [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+            onUpdate: {
+                extensionListModel.update(code)
+            }
+
+            onUninstall: {
+                extensionListModel.uninstall(code)
+            }
+
+            onOpenFullDescription: {
+//                extensionListModel.openFullDescription(code)
```
Comment (RomanPudashkin) [inline mu4/extensions/qml/MuseScore/Extensions/ExtensionsContent.qml:220]: Remove that comment or add todo, please

File: mu4/extensions/view/extensionlistmodel.cpp [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+    case rFileSize:
+        return QVariant::fromValue(item.fileSize);
+    case rStatus:
+        return QVariant::fromValue(static_cast<int>(item.status));
+    }
+
+    return QVariant();
+}
+
+int ExtensionListModel::rowCount(const QModelIndex& parent) const
```
Comment (RomanPudashkin) [inline mu4/extensions/view/extensionlistmodel.cpp:62]: int ExtensionListModel::rowCount(const QModelIndex&) const

File: mu4/scores/qml/MuseScore/Scores/NewScoreDialog.qml [hunk 20]
```diff
@@ -56,8 +56,6 @@ QmlDialog {
                 spacing: 20
 
                 FlatButton {
-
-                    width: 40
```
Comment (RomanPudashkin) [inline mu4/scores/qml/MuseScore/Scores/NewScoreDialog.qml:60]: Please revert all changes in NewScoreDialog.qml and ScoresModule.qml. We will get conflicts

File: framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+    id: root
+
+    property alias icon: buttonIcon.iconCode
+    property string text: ""
+    property string progressTitle: ""
+    property int iconPixelSize: buttonIcon.isEmpty ? 0 : 16
+
+    property real from: 0.0
+    property real to: 1.0
+    property real value: 0.0
```
Comment (RomanPudashkin) [inline framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml:14]: Can we use standard qml ProgressBar?

File: framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+    property bool indeterminate: false
+
+    signal clicked
```
Comment (RomanPudashkin) [inline framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml:18]: clicked()

File: mu4/appshell/qml/HomePage/AddonsContent.qml [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+
+        anchors.top: parent.top
+        anchors.topMargin: 66
+        anchors.left: parent.left
+        anchors.leftMargin: 133
+
+        font.pixelSize: 32
+        font.bold: true
+
+        text: qsTrc("appshell", "Add-ons")
```
Comment (RomanPudashkin) [inline mu4/appshell/qml/HomePage/AddonsContent.qml:26]: Why this component placed in appshell?

File: mu4/extensions/internal/extensionsconfiguration.cpp [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+
+QUrl ExtensionsConfiguration::extensionsUpdateUrl() const
+{
+    return QUrl("http://extensions.musescore.org/4.0/extensions/details.json");
+}
+
+QUrl ExtensionsConfiguration::extensionFileServerUrl(const QString &extensionCode) const
+{
+    ValCh<ExtensionsHash> _extensions = extensions();
+    QString fileName = _extensions.val.value(extensionCode).fileName;
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionsconfiguration.cpp:56]: Better to add function QString extensionFileName(extensionCode)

File: mu4/extensions/internal/extensionscontroller.cpp [hunk 25]
```diff
@@ -0,0 +1,369 @@
...
+
+    return make_ret(Err::NoError);
+}
+
+Extension::ExtensionTypes ExtensionsController::extensionTypes(const QString& extensionCode) const
+{
+    Extension::ExtensionTypes result;
+    QString workspacesPath(configuration()->extensionWorkspacesPath(extensionCode));
+    RetVal<QStringList> files
+        = fsOperation()->directoryFileList(workspacesPath, { QString("*.workspace") }, QDir::Files);
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionscontroller.cpp:305]: That line is duplicated (another one is in configuration)

File: mu4/extensions/internal/extensionunpacker.cpp [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+        return freeSpace;
+    }
+
+    RetVal2<QString, QVersionNumber> meta = extensionMeta(&zipFile);
+    if (!meta.ret) {
+        return meta.ret;
+    }
+
+    QString extensionId = meta.val1;
+    QString extensionDirPath(destination + "/" + extensionId);
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionunpacker.cpp:58]: Better to place this code in configuration

File: mu4/extensions/internal/extensionunpacker.cpp [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+                                          const QVersionNumber& version) const
+{
+    QDir destinationDir(destination);
+    QStringList dirList = destinationDir.entryList(QStringList(extensionId), QDir::Dirs | QDir::NoDotAndDotDot);
+
+    if (!dirList.contains(extensionId)) {
+        return make_ret(Err::NoError);
+    }
+
+    QString actualExtensionMetaFilePath = destination + "/" + extensionId + "/" + "metadata.json";
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionunpacker.cpp:130]: Better to place this code in configuration

Comment (Eism) [reply mu4/extensions/internal/extensionunpacker.cpp:130]: No, it is an independent class, it does not know about the other services.

File: framework/CMakeLists.txt (hunk 1 above)
Comment (Eism) [reply framework/CMakeLists.txt:9]: decided to merge these modules, remove the network module in another pull request.

File: mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml (hunk 15 above)
Comment (Eism) [reply mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml:36]: It is a placeholder, in the future we will change to a thumbnail extension.

File: mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml (hunk 17 above)
Comment (Eism) [reply mu4/extensions/qml/MuseScore/Extensions/ExtensionItem.qml]: yes, thanks

File: framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml (hunk 21 above)
Comment (Eism) [reply framework/uicomponents/qml/MuseScore/UiComponents/ProgressButton.qml:15]: No, this is problem for this task.

File: mu4/extensions/internal/extensionsconfiguration.h [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace mu {
+namespace extensions {
+class ExtensionsConfiguration : public IExtensionsConfiguration
+{
+    INJECT(extensions, framework::IGlobalConfiguration, globalConfiguration)
+    INJECT(extensions, framework::IFsOperations, fsOperation)
+
+public:
+    ExtensionsConfiguration() = default;
```
Comment (RomanPudashkin) [inline mu4/extensions/internal/extensionsconfiguration.h:35]: Will be generated automatically

**Pull request merged**